File: src/main/java/no/nav/tag/dittNavArbeidsgiver/clients/altinn/AltinnTilgangssøknadClient.java
package no.nav.tag.dittNavArbeidsgiver.clients.altinn;

import lombok.extern.slf4j.Slf4j;
import no.nav.tag.dittNavArbeidsgiver.clients.altinn.dto.DelegationRequest;
import no.nav.tag.dittNavArbeidsgiver.clients.altinn.dto.Søknadsstatus;
import no.nav.tag.dittNavArbeidsgiver.models.AltinnTilgangssøknad;
import no.nav.tag.dittNavArbeidsgiver.models.AltinnTilgangssøknadsskjema;
import no.nav.tag.dittNavArbeidsgiver.services.altinn.AltinnConfig;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.RequestEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;
import java.util.stream.Collectors;

@Slf4j
@Component
public class AltinnTilgangssøknadClient {
    private final RestTemplate restTemplate;
    private final HttpHeaders altinnHeaders;
    private final Supplier<UriComponentsBuilder> altinnUriBuilder;
    private final URI altinnUri;
    private final ParameterizedTypeReference<Søknadsstatus> søknadsstatusType;
    private final ParameterizedTypeReference<DelegationRequest> delegationRequestType;

    @Autowired
    public AltinnTilgangssøknadClient(RestTemplate restTemplate, AltinnConfig altinnConfig) {
        this.restTemplate = restTemplate;

        var altinnUriBuilder = UriComponentsBuilder.fromUriString(
                altinnConfig.getProxyFallbackUrl()
                        + "/ekstern/altinn/api/serviceowner/delegationRequests?ForceEIAuthentication"
        );

        this.altinnUriBuilder = altinnUriBuilder::cloneBuilder;


        this.altinnUri = altinnUriBuilder.build().toUri();
        log.info("altinnUri: {}", altinnUri);

        this.altinnHeaders = new HttpHeaders();
        this.altinnHeaders.add("accept", "application/hal+json");
        this.altinnHeaders.add("apikey", altinnConfig.getAltinnHeader());
        this.altinnHeaders.add("x-nav-apikey", altinnConfig.getAPIGwHeader());

        this.søknadsstatusType = new ParameterizedTypeReference<>() {
        };

        this.delegationRequestType = new ParameterizedTypeReference<>() {
        };
    }


    public List<AltinnTilgangssøknad> hentSøknader(String fødselsnummer) {
        var resultat = new ArrayList<AltinnTilgangssøknad>();
        String filter = String.format("CoveredBy eq '%s'", fødselsnummer);

        URI uri = altinnUriBuilder.get()
                .queryParam("$filter", filter)
                .build()
                .toUri();
        while (uri != null) {
            var request = RequestEntity.get(uri).headers(altinnHeaders).build();
            var response = restTemplate.exchange(request, søknadsstatusType);

            if (response.getStatusCode() != HttpStatus.OK) {
                var msg = String.format("Henting av status på tilgangssøknader feilet med http-status %s", response.getStatusCode());
                log.error(msg);
                throw new RuntimeException(msg);
            }

            var body = response.getBody();
            if (body == null) {
                log.warn("Altinn delegation requests: body missing");
                break;
            }
            if (body.embedded.delegationRequests.isEmpty()) {
                uri = null;
            } else {
                uri = altinnUriBuilder
                        .get()
                        .queryParam("$filter", filter)
                        .queryParam("continuation", body.continuationtoken)
                        .build()
                        .toUri();
            }

            body.embedded
                    .delegationRequests
                    .stream()
                    .map(søknadDTO -> {
                        var søknad = new AltinnTilgangssøknad();
                        søknad.setOrgnr(søknadDTO.OfferedBy);
                        søknad.setStatus(søknadDTO.RequestStatus);
                        søknad.setCreatedDateTime(søknadDTO.Created);
                        søknad.setLastChangedDateTime(søknadDTO.LastChanged);
                        søknad.setServiceCode(søknadDTO.RequestResources.get(0).ServiceCode);
                        søknad.setServiceEdition(søknadDTO.RequestResources.get(0).ServiceEditionCode);
                        søknad.setSubmitUrl(søknadDTO.links.sendRequest.href);
                        return søknad;
                    })
                    .collect(Collectors.toCollection(() -> resultat));
        }

        return resultat;
    }

    public AltinnTilgangssøknad sendSøknad(String fødselsnummer, AltinnTilgangssøknadsskjema søknadsskjema) {
        var requestResource = new DelegationRequest.RequestResource();
        requestResource.ServiceCode = søknadsskjema.serviceCode;
        requestResource.ServiceEditionCode = søknadsskjema.serviceEdition;

        var delegationRequest = new DelegationRequest();
        delegationRequest.CoveredBy = fødselsnummer;
        delegationRequest.OfferedBy = søknadsskjema.orgnr;
        delegationRequest.RedirectUrl = søknadsskjema.redirectUrl;
        delegationRequest.RequestResources = List.of(requestResource);

        var request = RequestEntity
                .post(altinnUri)
                .headers(altinnHeaders)
                .body(delegationRequest);

        var response = restTemplate.exchange(request, delegationRequestType);

        if (response.getStatusCode() != HttpStatus.OK) {
            var msg = String.format("Ny tilgangssøknad i altinn feilet med http-status %s", response.getStatusCode());
            log.error(msg);
            throw new RuntimeException(msg);
        }

        var body = response.getBody();

        var svar = new AltinnTilgangssøknad();
        svar.setStatus(body.RequestStatus);
        svar.setSubmitUrl(body.links.sendRequest.href);
        return svar;
    }
}

File: src/main/kotlin/no/nav/arbeidsgiver/min_side/tilgangssoknad/AltinnTilgangssøknadClient.kt
package no.nav.arbeidsgiver.min_side.tilgangssoknad

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import no.nav.arbeidsgiver.min_side.config.logger
import no.nav.arbeidsgiver.min_side.config.retryInterceptor
import no.nav.arbeidsgiver.min_side.maskinporten.MaskinportenTokenService
import no.nav.arbeidsgiver.min_side.tilgangssoknad.DelegationRequest.RequestResource
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.web.client.RestTemplateBuilder
import org.springframework.http.HttpHeaders
import org.springframework.http.RequestEntity
import org.springframework.stereotype.Component
import org.springframework.web.client.HttpClientErrorException.BadRequest
import org.springframework.web.client.HttpServerErrorException
import org.springframework.web.client.ResourceAccessException
import org.springframework.web.client.exchange
import org.springframework.web.util.UriComponentsBuilder
import java.net.SocketException
import javax.net.ssl.SSLHandshakeException

@Suppress("UastIncorrectHttpHeaderInspection")
@Component
class AltinnTilgangssøknadClient(
    restTemplateBuilder: RestTemplateBuilder,
    @Value("\${altinn.apiBaseUrl}") altinnApiBaseUrl: String,
    @Value("\${altinn.altinnHeader}") private val altinnApiKey: String,
    private val maskinportenTokenService: MaskinportenTokenService,
) {
    private val log = logger()

    private val restTemplate = restTemplateBuilder
        .rootUri(altinnApiBaseUrl)
        .additionalInterceptors(
            retryInterceptor(
                maxAttempts = 3,
                backoffPeriod = 250L,
                HttpServerErrorException.BadGateway::class.java,
                HttpServerErrorException.ServiceUnavailable::class.java,
                HttpServerErrorException.GatewayTimeout::class.java,
                SocketException::class.java,
                SSLHandshakeException::class.java,
                ResourceAccessException::class.java,
            )
        )
        .build()

    private val delegationRequestApiPath = UriComponentsBuilder
        .fromUriString(altinnApiBaseUrl)
        .path("/api/serviceowner/delegationRequests")
        .build()
        .toUriString()

    fun hentSøknader(fødselsnummer: String): List<AltinnTilgangssøknad> {
        val filter = "CoveredBy eq '$fødselsnummer'"
        var shouldContinue = true
        var continuationtoken: String? = null
        val resultat = ArrayList<AltinnTilgangssøknad>()
        while (shouldContinue) {
            val uri =
                "$delegationRequestApiPath?ForceEIAuthentication&${
                    if (continuationtoken == null) {
                        "\$filter={filter}"
                    } else {
                        "\$filter={filter}&continuation={continuation}"
                    }
                }"

            val body = try {
                restTemplate.exchange<Søknadsstatus?>(
                    RequestEntity.get(uri, filter, continuationtoken)
                        .headers(HttpHeaders().apply {
                            set("accept", "application/hal+json")
                            set("apikey", altinnApiKey)
                            setBearerAuth(maskinportenTokenService.currentAccessToken())
                        }).build()
                ).body
            } catch (e: BadRequest) {
                if (e.message!!.contains("User profile")) { // Altinn returns 400 if user does not exist
                    null
                } else {
                    throw e
                }
            }

            if (body == null) {
                log.error("Altinn delegation requests: body missing")
                break
            }

            if (body.embedded!!.delegationRequests!!.isEmpty()) {
                shouldContinue = false
            } else {
                continuationtoken = body.continuationtoken
            }

            body.embedded.delegationRequests!!.mapTo(resultat) { søknadDTO: DelegationRequest ->
                AltinnTilgangssøknad(
                    orgnr = søknadDTO.OfferedBy,
                    status = søknadDTO.RequestStatus,
                    createdDateTime = søknadDTO.Created,
                    lastChangedDateTime = søknadDTO.LastChanged,
                    serviceCode = søknadDTO.RequestResources!![0].ServiceCode,
                    serviceEdition = søknadDTO.RequestResources[0].ServiceEditionCode,
                    submitUrl = søknadDTO.links!!.sendRequest!!.href,
                )
            }
        }
        return resultat
    }

    fun sendSøknad(fødselsnummer: String, søknadsskjema: AltinnTilgangssøknadsskjema): AltinnTilgangssøknad {
        val response = restTemplate.exchange<DelegationRequest?>(
            RequestEntity
                .post("$delegationRequestApiPath?ForceEIAuthentication")
                .headers(HttpHeaders().apply {
                    set("accept", "application/hal+json")
                    set("apikey", altinnApiKey)
                    setBearerAuth(maskinportenTokenService.currentAccessToken())
                })
                .body(
                    DelegationRequest(
                        CoveredBy = fødselsnummer,
                        OfferedBy = søknadsskjema.orgnr,
                        RedirectUrl = søknadsskjema.redirectUrl,
                        RequestResources = listOf(
                            RequestResource(
                                ServiceCode = søknadsskjema.serviceCode,
                                ServiceEditionCode = søknadsskjema.serviceEdition,
                            )
                        ),
                    )
                )
        )
        return response.body.let {
            AltinnTilgangssøknad(
                status = it!!.RequestStatus,
                submitUrl = it.links!!.sendRequest!!.href,
            )
        }
    }
}

@Suppress("PropertyName")
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
data class DelegationRequest(
    val RequestStatus: String? = null,
    val OfferedBy: String? = null,
    val CoveredBy: String? = null,
    val RedirectUrl: String? = null,
    val Created: String? = null,
    val LastChanged: String? = null,
    val KeepSessionAlive: Boolean = true,
    val RequestResources: List<RequestResource>? = null,
    @field:JsonProperty("_links") val links: Links? = null
) {

    @JsonIgnoreProperties(ignoreUnknown = true)
    data class RequestResource(
        val ServiceCode: String? = null,
        val ServiceEditionCode: Int? = null
    )

    @JsonIgnoreProperties(ignoreUnknown = true)
    data class Links(var sendRequest: Link? = null)

    @JsonIgnoreProperties(ignoreUnknown = true)
    data class Link(var href: String? = null)
}

@JsonIgnoreProperties(ignoreUnknown = true)
data class Søknadsstatus(
    @field:JsonProperty("_embedded")
    val embedded: Embedded? = null,
    val continuationtoken: String? = null,
) {
    @JsonIgnoreProperties(ignoreUnknown = true)
    data class Embedded(
        val delegationRequests: List<DelegationRequest>? = null
    )
}

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
data class AltinnTilgangssøknad(
    val orgnr: String? = null,
    val serviceCode: String? = null,
    val serviceEdition: Int? = null,
    val status: String? = null,
    val createdDateTime: String? = null,
    val lastChangedDateTime: String? = null,
    val submitUrl: String? = null
)

data class AltinnTilgangssøknadsskjema(
    val orgnr: String,
    val redirectUrl: String,
    val serviceCode: String,
    val serviceEdition: Int,
)